File: py/gateway/cloud_api.py
"""Cloud API evidence retrieval and response decoding.

These helpers fetch or select evidence only. The caller must pass the result
to ``verify_gateway_attestation`` or ``verify_model_attestation`` before using
it as verified evidence.
"""

from __future__ import annotations

import http.client
import json
import os
import secrets
import ssl
from hashlib import sha256
from typing import Any, Dict, List, Optional, Tuple
from urllib.parse import urlencode, urlsplit

import requests
from cryptography import x509
from cryptography.hazmat.primitives import serialization

from py.common.dstack_attestation import (
    signing_identities_match,
    signing_identity,
)


BASE_URL = os.environ.get("BASE_URL", "https://cloud-api.near.ai")
API_KEY = os.environ.get("API_KEY", "")


def cloud_api_url(path: str) -> str:
    """Return one Cloud API URL while accepting a base URL ending in ``/v1``."""

    base = BASE_URL.rstrip("/")
    if base.endswith("/v1"):
        return f"{base}/{path.lstrip('/')}"
    return f"{base}/v1/{path.lstrip('/')}"


def cloud_api_headers(extra: Optional[Dict[str, str]] = None) -> Dict[str, str]:
    """Return request headers with the configured API key when present."""

    headers = {} if extra is None else dict(extra)
    if API_KEY:
        headers["Authorization"] = f"Bearer {API_KEY}"
    return headers


def _request_json(
    path: str,
    params: Dict[str, str],
    headers: Optional[Dict[str, str]] = None,
) -> Dict[str, Any]:
    response = requests.get(
        cloud_api_url(path),
        params=params,
        headers=cloud_api_headers(headers),
        timeout=30,
    )
    response.raise_for_status()
    value = response.json()
    if not isinstance(value, dict):
        raise ValueError(f"{path} returned {type(value).__name__}, expected a JSON object")
    return value


def fetch_model_attestations(
    model: str,
    signing_algo: Optional[str] = None,
    signing_address: Optional[str] = None,
) -> Tuple[List[Dict[str, Any]], str]:
    """Fetch NEAR model evidence with a fresh nonce and no TLS binding."""

    nonce = secrets.token_hex(32)
    params = {
        "model": model,
        "provider": "near",
        "nonce": nonce,
        "include_tls_fingerprint": "false",
    }
    if signing_algo is not None:
        params["signing_algo"] = signing_algo
    if signing_address is not None:
        params["signing_address"] = signing_address
    report = _request_json(
        "attestation/report",
        params,
        headers={"x-no-aliasing": "true"},
    )
    attestations = report.get("model_attestations")
    if not isinstance(attestations, list) or not all(
        isinstance(attestation, dict) for attestation in attestations
    ):
        raise ValueError("Cloud API model report does not contain model_attestations[]")
    return attestations, nonce


def _peer_spki_fingerprint(certificate_der: bytes) -> str:
    certificate = x509.load_der_x509_certificate(certificate_der)
    spki_der = certificate.public_key().public_bytes(
        serialization.Encoding.DER,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return sha256(spki_der).hexdigest()


def _fetch_gateway_report_with_peer_spki(
    params: Dict[str, str]
) -> Tuple[Dict[str, Any], Optional[str]]:
    """Fetch Gateway evidence and retain the TLS peer from that same request."""

    url = urlsplit(cloud_api_url("attestation/report"))
    if url.scheme != "https" or not url.hostname:
        # Local HTTP is useful for development but has no peer certificate.
        return _request_json("attestation/report", params), None

    path = url.path or "/"
    query = urlencode(params)
    if url.query:
        query = f"{url.query}&{query}"
    if query:
        path = f"{path}?{query}"

    connection = http.client.HTTPSConnection(
        url.hostname,
        url.port or 443,
        context=ssl.create_default_context(),
        timeout=30,
    )
    try:
        connection.connect()
        socket = connection.sock
        if socket is None:
            raise RuntimeError("HTTPS connection did not expose a TLS socket")
        certificate_der = socket.getpeercert(binary_form=True)
        if not certificate_der:
            raise RuntimeError("TLS peer did not provide a certificate")
        peer_spki = _peer_spki_fingerprint(certificate_der)

        connection.request("GET", path, headers=cloud_api_headers())
        response = connection.getresponse()
        body = response.read()
        if not 200 <= response.status < 300:
            rendered = body.decode("utf-8", "replace")
            raise RuntimeError(f"Cloud API returned HTTP {response.status}: {rendered}")
        value = json.loads(body)
        if not isinstance(value, dict):
            raise ValueError("attestation/report returned a non-object JSON value")
        return value, peer_spki
    finally:
        connection.close()


def fetch_gateway_attestation(
    signing_algo: Optional[str] = None,
) -> Tuple[Dict[str, Any], str, Optional[str]]:
    """Fetch standalone Gateway evidence and the TLS peer that served it."""

    nonce = secrets.token_hex(32)
    params = {"nonce": nonce, "include_tls_fingerprint": "true"}
    if signing_algo is not None:
        params["signing_algo"] = signing_algo
    report, peer_spki = _fetch_gateway_report_with_peer_spki(params)
    attestation = report.get("gateway_attestation")
    if not isinstance(attestation, dict):
        raise ValueError("Cloud API Gateway report does not contain gateway_attestation")
    return attestation, nonce, peer_spki


def fetch_model_attestation_for_signature(
    model: str,
    signature: Dict[str, Any],
) -> Tuple[Dict[str, Any], str]:
    """Fetch exactly one model attestation for a ``provider_tee`` signature."""

    if signature.get("signature_kind") != "provider_tee":
        raise ValueError("Model evidence can only verify a provider_tee signature")
    signing_algo, _ = signing_identity(signature, "signature")
    signing_address = signature["signing_address"]
    attestations, nonce = fetch_model_attestations(
        model,
        signing_algo=signing_algo,
        signing_address=signing_address,
    )
    matches = [
        attestation
        for attestation in attestations
        if signing_identities_match(attestation, signature)
    ]
    if len(matches) != 1:
        raise ValueError(
            "Cloud API must return exactly one model attestation matching the signature signer; "
            f"received {len(matches)} matching entries"
        )
    return matches[0], nonce


def fetch_gateway_attestation_for_signature(
    signature: Dict[str, Any],
) -> Tuple[Dict[str, Any], str, Optional[str]]:
    """Fetch Gateway evidence for a ``gateway`` signature."""

    if signature.get("signature_kind") != "gateway":
        raise ValueError("Gateway evidence can only verify a gateway signature")
    signing_algo, _ = signing_identity(signature, "signature")
    return fetch_gateway_attestation(signing_algo=signing_algo)

File: ts/gateway/cloud_api.ts
/**
 * Cloud API evidence retrieval and response decoding.
 *
 * This module fetches evidence only. It does not turn a response into trusted
 * evidence; callers must pass the result to the matching attestation verifier.
 */

import { Buffer } from 'buffer';
import * as crypto from 'crypto';
import * as https from 'https';
import * as tls from 'tls';
import { URL } from 'url';

import type { CompletionSignature } from './completion';
import {
  type AttestationReport,
  type SigningAlgo,
  normalizeSigningAlgo,
  signingAddressBytes,
  verifyAttestationNonce,
} from '../common/dstack_attestation';

export type { AttestationReport, SigningAlgo };

export interface AttestationApiReport {
  gateway_attestation?: AttestationReport;
  model_attestations?: AttestationReport[];
  tls_certificate?: string;
  [key: string]: unknown;
}

export interface FetchModelAttestationsParams {
  model: string;
  nonce: string;
  signingAlgo?: SigningAlgo;
  signingAddress?: string;
}

export interface FindModelAttestationForSignatureParams {
  attestations: readonly AttestationReport[];
  signature: CompletionSignature;
}

export interface FetchGatewayAttestationParams {
  nonce: string;
  signingAlgo?: SigningAlgo;
}

export interface FetchedGatewayAttestation {
  attestation: AttestationReport;
  peerSpkiFingerprint?: string;
}

type JsonRecord = Record<string, unknown>;

interface ParseAttestationParams {
  value: unknown;
  label: string;
  requireReportData?: boolean;
}

function cloudApiBaseUrl(): string {
  return process.env.BASE_URL || 'https://cloud-api.near.ai';
}

function cloudApiHeaders(extra: Record<string, string> = {}): Record<string, string> {
  const apiKey = process.env.API_KEY;
  return {
    'Accept-Encoding': 'identity',
    ...(apiKey ? { Authorization: `Bearer ${apiKey}` } : {}),
    ...extra,
  };
}

function attestationReportUrl(query: Record<string, string>): URL {
  const url = new URL('/v1/attestation/report', cloudApiBaseUrl());
  for (const [name, value] of Object.entries(query)) {
    url.searchParams.set(name, value);
  }
  return url;
}

function asRecord(value: unknown, label: string): JsonRecord {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`${label} must be a JSON object`);
  }
  return value as JsonRecord;
}

function requireString(value: unknown, label: string): string {
  if (typeof value !== 'string' || value.length === 0) {
    throw new Error(`${label} must be a non-empty string`);
  }
  return value;
}

async function requestJson(
  url: URL,
  headers: Record<string, string>,
): Promise<unknown> {
  const response = await fetch(url, { headers });
  const body = await response.text();
  if (!response.ok) {
    throw new Error(`Cloud API returned HTTP ${response.status}: ${body}`);
  }
  try {
    return JSON.parse(body);
  } catch (cause) {
    throw new Error(
      `Cloud API returned invalid JSON: ${cause instanceof Error ? cause.message : String(cause)}`,
    );
  }
}

/**
 * Fetch JSON and capture the SPKI fingerprint of the TLS peer that served
 * that evidence response. This is intentionally Node-only: browsers do not
 * expose a peer certificate to JavaScript.
 */
function requestJsonWithPeerSpki(
  url: URL,
  headers: Record<string, string>,
): Promise<{ body: unknown; peerSpkiFingerprint?: string }> {
  if (url.protocol !== 'https:') {
    return requestJson(url, headers).then((body) => ({ body }));
  }

  return new Promise((resolve, reject) => {
    const request = https.request(
      {
        protocol: url.protocol,
        hostname: url.hostname,
        port: url.port || 443,
        path: `${url.pathname}${url.search}`,
        method: 'GET',
        headers,
        // This is the connection whose TLS peer key is compared against the
        // quote. Do not take it from a reused keep-alive agent.
        agent: false,
        timeout: 30_000,
      },
      (response) => {
        let peerSpkiFingerprint: string | undefined;
        try {
          const socket = response.socket as tls.TLSSocket;
          const certificate = peerCertificate(socket);
          if (certificate) {
            const spkiDer = certificate.publicKey.export({
              type: 'spki',
              format: 'der',
            });
            peerSpkiFingerprint = crypto
              .createHash('sha256')
              .update(spkiDer)
              .digest('hex');
          }
        } catch (cause) {
          reject(
            new Error(
              `Could not read the TLS peer certificate: ${
                cause instanceof Error ? cause.message : String(cause)
              }`,
            ),
          );
          response.resume();
          return;
        }

        const chunks: Buffer[] = [];
        response.on('data', (chunk: Buffer) => chunks.push(Buffer.from(chunk)));
        response.on('error', reject);
        response.on('end', () => {
          const text = Buffer.concat(chunks).toString('utf8');
          if ((response.statusCode ?? 500) < 200 || (response.statusCode ?? 500) >= 300) {
            reject(new Error(`Cloud API returned HTTP ${response.statusCode}: ${text}`));
            return;
          }
          try {
            resolve({ body: JSON.parse(text), peerSpkiFingerprint });
          } catch (cause) {
            reject(
              new Error(
                `Cloud API returned invalid JSON: ${
                  cause instanceof Error ? cause.message : String(cause)
                }`,
              ),
            );
          }
        });
      },
    );
    request.on('error', reject);
    request.on('timeout', () => {
      request.destroy(new Error('Cloud API request timed out'));
    });
    request.end();
  });
}

/**
 * Node does not always retain an X509Certificate object for a live TLS
 * socket. The raw peer certificate contains the same SPKI material.
 */
function peerCertificate(socket: tls.TLSSocket): crypto.X509Certificate | undefined {
  const certificate = socket.getPeerX509Certificate();
  if (certificate !== undefined) {
    return certificate;
  }

  const peer = socket.getPeerCertificate(true);
  return peer.raw === undefined ? undefined : new crypto.X509Certificate(peer.raw);
}

function parseAttestation({
  value,
  label,
  requireReportData = false,
}: ParseAttestationParams): AttestationReport {
  const record = asRecord(value, label);
  const info = asRecord(record.info, `${label}.info`);
  if (
    record.nvidia_payload !== undefined &&
    record.nvidia_payload !== null &&
    typeof record.nvidia_payload !== 'string'
  ) {
    throw new Error(`${label}.nvidia_payload must be a string or null`);
  }
  if (
    record.tls_cert_fingerprint !== undefined &&
    record.tls_cert_fingerprint !== null &&
    typeof record.tls_cert_fingerprint !== 'string'
  ) {
    throw new Error(`${label}.tls_cert_fingerprint must be a string or null`);
  }
  if (requireReportData && typeof record.report_data !== 'string') {
    throw new Error(`${label}.report_data must be a string`);
  }
  if (
    record.report_data !== undefined &&
    record.report_data !== null &&
    typeof record.report_data !== 'string'
  ) {
    throw new Error(`${label}.report_data must be a string or null`);
  }
  return {
    intel_quote: requireString(record.intel_quote, `${label}.intel_quote`),
    signing_address: requireString(record.signing_address, `${label}.signing_address`),
    signing_algo: requireString(record.signing_algo, `${label}.signing_algo`),
    info: {
      tcb_info: info.tcb_info as AttestationReport['info']['tcb_info'],
    },
    ...(typeof record.event_log === 'string' || Array.isArray(record.event_log)
      ? { event_log: record.event_log }
      : {}),
    ...(typeof record.report_data === 'string' ? { report_data: record.report_data } : {}),
    ...(typeof record.request_nonce === 'string'
      ? { request_nonce: record.request_nonce }
      : {}),
    ...(record.nvidia_payload === null || typeof record.nvidia_payload === 'string'
      ? { nvidia_payload: record.nvidia_payload }
      : {}),
    ...(record.tls_cert_fingerprint === null || typeof record.tls_cert_fingerprint === 'string'
      ? { tls_cert_fingerprint: record.tls_cert_fingerprint }
      : {}),
    ...(typeof record.model_name === 'string' ? { model_name: record.model_name } : {}),
  };
}

function parseAttestationApiReport(value: unknown): AttestationApiReport {
  const record = asRecord(value, 'Cloud API attestation response');
  const report: AttestationApiReport = {};
  if (record.gateway_attestation !== undefined) {
    report.gateway_attestation = parseAttestation({
      value: record.gateway_attestation,
      label: 'gateway_attestation',
      requireReportData: true,
    });
  }
  if (record.model_attestations !== undefined) {
    if (!Array.isArray(record.model_attestations)) {
      throw new Error('model_attestations must be an array when present');
    }
    report.model_attestations = record.model_attestations.map((item, index) =>
      parseAttestation({
        value: item,
        label: `model_attestations[${index}]`,
      }),
    );
  }
  if (typeof record.tls_certificate === 'string') {
    report.tls_certificate = record.tls_certificate;
  }
  return report;
}

/**
 * Fetch every NEAR model evidence item returned for a standalone deployment
 * audit. Optional signer filters only narrow the API response; this does not
 * associate the evidence with a completion signature.
 */
export async function fetchModelAttestations({
  model,
  nonce,
  signingAlgo,
  signingAddress,
}: FetchModelAttestationsParams): Promise<AttestationReport[]> {
  const query: Record<string, string> = {
    model,
    nonce,
    provider: 'near',
    include_tls_fingerprint: 'false',
  };
  if (signingAlgo !== undefined) query.signing_algo = signingAlgo;
  if (signingAddress !== undefined) query.signing_address = signingAddress;
  const report = parseAttestationApiReport(
    await requestJson(
      attestationReportUrl(query),
      cloudApiHeaders({ 'x-no-aliasing': 'true' }),
    ),
  );
  const attestations = report.model_attestations ?? [];
  if (attestations.length === 0) {
    throw new Error('Cloud API response does not contain model_attestations');
  }
  for (const attestation of attestations) {
    verifyAttestationNonce(attestation, nonce);
  }
  return attestations;
}

/**
 * Select the one NEAR model evidence item whose advertised signer matches a
 * provider-TEE signature. This does not verify the quote or the signature.
 */
export function findModelAttestationForSignature({
  attestations,
  signature,
}: FindModelAttestationForSignatureParams): AttestationReport {
  if (signature.kind !== 'provider_tee') {
    throw new Error('Model evidence can only verify a provider_tee signature');
  }

  const candidates = attestations.filter((attestation) => {
    try {
      return (
        normalizeSigningAlgo(attestation.signing_algo) === signature.signingAlgo &&
        signingAddressBytes(
          attestation.signing_address,
          signature.signingAlgo,
        ).equals(
          signingAddressBytes(signature.signingAddress, signature.signingAlgo),
        )
      );
    } catch {
      return false;
    }
  });
  if (candidates.length !== 1) {
    throw new Error(
      `Expected exactly one NEAR model attestation for the signature signer; found ${candidates.length}`,
    );
  }
  return candidates[0]!;
}

/**
 * Fetch Gateway evidence with TLS binding enabled. The peer fingerprint comes
 * from the HTTPS connection that served this evidence response, not a
 * completion request.
 */
export async function fetchGatewayAttestation({
  nonce,
  signingAlgo,
}: FetchGatewayAttestationParams): Promise<FetchedGatewayAttestation> {
  const query: Record<string, string> = {
    nonce,
    include_tls_fingerprint: 'true',
  };
  if (signingAlgo !== undefined) query.signing_algo = signingAlgo;
  const response = await requestJsonWithPeerSpki(
    attestationReportUrl(query),
    cloudApiHeaders(),
  );
  const report = parseAttestationApiReport(response.body);
  if (!report.gateway_attestation) {
    throw new Error('Cloud API response does not contain gateway_attestation');
  }
  verifyAttestationNonce(report.gateway_attestation, nonce);
  return {
    attestation: report.gateway_attestation,
    ...(response.peerSpkiFingerprint
      ? { peerSpkiFingerprint: response.peerSpkiFingerprint }
      : {}),
  };
}
